promq/autocomplete/earley/completer.go

```diff
@@ -107,86 +107,40 @@ func (c *promQLCompleter) GenerateSuggestions(query string, pos int) []autocompl
-			}
-		case COMPARISION:
-			for _, ao := range autocomplete.FilterPrefix(sets.StringKeySet(comparisionOperators), autocompletePrefix, false).List() {
-				newMatch := NewPartialMatch(ao, "comparision", comparisionOperators[ao])
-				matches = append(matches, newMatch)
-			}
-		case SET:
-			for _, ao := range autocomplete.FilterPrefix(sets.StringKeySet(setOperators), autocompletePrefix, false).List() {
-				newMatch := NewPartialMatch(ao, "logical", setOperators[ao])
-				matches = append(matches, newMatch)
-			}
-		case LABELMATCH:
-			for _, ao := range autocomplete.FilterPrefix(sets.StringKeySet(labelMatchOperators), autocompletePrefix, false).List() {
-				newMatch := NewPartialMatch(ao, "label-match", labelMatchOperators[ao])
-				matches = append(matches, newMatch)
-			}
-		case UNARY_OP:
-			for _, ao := range autocomplete.FilterPrefix(sets.StringKeySet(unaryOperators), autocompletePrefix, false).List() {
-				newMatch := NewPartialMatch(ao, "unary", unaryOperators[ao])
-				matches = append(matches, newMatch)
-			}
-		case OFFSET_KW:
-			matches = append(matches, NewPartialMatch("offset", "keyword", keywords["offset"]))
-		case BOOL_KW:
-			matches = append(matches, NewPartialMatch("bool", "keyword", keywords["bool"]))
-		case GROUP_SIDE:
-			for _, ao := range autocomplete.FilterPrefix(sets.StringKeySet(groupSideKeywords), autocompletePrefix, false).List() {
-				newMatch := NewPartialMatch(ao, "keyword", groupSideKeywords[ao])
-				matches = append(matches, newMatch)
+		case s.TokenType == DURATION:
+			// add time units to match is the prefix is number
+			if autocompletePrefix == "" {
+				continue
 			}
-		case GROUP_KW:
-			for _, ao := range autocomplete.FilterPrefix(sets.StringKeySet(groupKeywords), autocompletePrefix, false).List() {
-				newMatch := NewPartialMatch(ao, "keyword", groupKeywords[ao])
-				matches = append(matches, newMatch)
+			// subquery expression has range and resolution that are split by ":"
+			if strings.Contains(autocompletePrefix, ":") {
+				autocompletePrefix = strings.Split(autocompletePrefix, ":")[1]
 			}
-		case DURATION:
-			// add time units to match is the prefix is number
 			if _, err := strconv.Atoi(autocompletePrefix); err == nil {
 				for _, ao := range sets.StringKeySet(timeUnits).List() {
 					newMatch := NewPartialMatch(ao, "time-unit", timeUnits[ao])
 					matches = append(matches, newMatch)
 				}
 			}
-		case FUNCTION_VECTOR_ID:
-			for _, ao := range autocomplete.FilterPrefix(sets.StringKeySet(vectorFunctions), autocompletePrefix, false).List() {
-				newMatch := NewPartialMatch(ao, "label-match", vectorFunctions[ao])
-				matches = append(matches, newMatch)
-			}
-		case FUNCTION_SCALAR_ID:
-			for _, ao := range autocomplete.FilterPrefix(sets.StringKeySet(scalarFunctions), autocompletePrefix, false).List() {
-				newMatch := NewPartialMatch(ao, "label-match", scalarFunctions[ao])
+		case newStringSet(tokenTypes...).Has(string(s.TokenType)):
```

Comment:
isn't newStringSet(tokenTypes...) basically a constant? shouldn't we just do this once and cache it?

Reply:
Oh yes. Added var `tokenTypeStringSet` in promql.go